Automattic/newspack-plugin
# fix(ras): consolidate RAS <> ESP data syncs in data event connectors

### All Submissions:






### Changes proposed in this Pull Request:



#2804 fixed syncing of contact metadata for UTM data via purchases, but did not fix for registration via non-transactions. This PR fixes syncing of UTM fields for all registration and purchase-related events, and also does a bit of refactoring to consolidate data syncs in general into the ActiveCampaign and Mailchimp Data Events connectors. Currently, the data syncs are happening in a few different places using different triggers (a mix of Woo hooks, regular WP hooks, Newspack Newsletter hooks, and data events throughout different files). Consolidating them in data events should make it easier to see and control what triggers an ESP sync going forward. As part of this we've added some new data event listeners and tweaked some existing ones to make them a bit more standardized.

### How to test the changes in this Pull Request:

In new incognito sessions, perform the following actions from a URL with UTM parameters appended (e.g. `?utm_source=marketinglist&utm_medium=email&utm_campaign=2023-Newspack-Campaign`):

- Register for a new account via the Registration block without signing up for any newsletter lists
- Sign up for newsletter lists via the Subscription Form block or the Registration block
- Complete a one-time donation
- Complete a subscription donation
- Complete a non-donation non-subscription purchase
- Complete a non-donation subscription purchase

Make sure to repeat the above for both Mailchimp and ActiveCampaign (so the end result should be 12 new contacts). In each case, find the contact synced to the ESP and confirm that UTM metadata fields are synced. For registration-based actions, the field keys should start with `Signup UTM:`. For payment-based actions, the field keys should start with `Payment UTM:`.

Please also smoke test that logging into an existing account and making a renewal order on an active subscription also continue to trigger an ESP data sync as before.

### Other information:




---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206082237141820
